RevisionReader: Parse raw commit body

gerhardol opened the pull request:
Part of #7635 and #9229
                                      
Parse raw body %B instead of combining %s%b to avoid extra processing
in both Git and GE.

## Proposed changes

See #7635 and #9229 for background.
This decreases the time to list Git revisions with a few percent, as well as cleaning up the code and tests.

## Test methodology <!-- How did you ensure quality? -->

Tests were adapted

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie reviewed (commented):
👍

drewnoakes commented on GitCommands/RevisionReader.cs:
```diff
@@ -448,32 +445,40 @@ DateTime ParseUnixDateTime()
-            string tail = reader.ReadToEnd() ?? "";
-            int indexOfEndOfBody = tail.LastIndexOf(EndOfBody, StringComparison.InvariantCulture);
+            // Ignore leading empty or whitespace lines, see Git pretty.c
+            string tail = reader.ReadToEnd().TrimStart() ?? "";
```
Is there a way to prevent allocating this tail string? Perhaps by extending the reader? For large repositories, this may create a lot of objects for the GC.

Eg:

reader.SkipWhiteSpace();
var body = reader.ReadTo(EndOfBody);

gerhardol replied:
For end of body, I would like to remove that with a single null, and read additional data separately. Not finished yet. 

This implementation should have the same allocations as the current.

gerhardol replied:
Changing to something like the suggested to simplify further, will simplify updated parsing of the reader. 

ParseCommitBody() is no longer needed, removing it and the tests for it. 
This may be be replaced with parsing the git-log output instead eventually.

drewnoakes replied:
Nice.

drewnoakes commented on GitCommands/RevisionReader.cs:
```diff
@@ -52,8 +52,7 @@ public sealed class RevisionReader : IDisposable
               /* Committer email */ "%cE%n" +
-              /* Commit subject  */ "%s%n%n" +
-              /* Commit body     */ "%b" + EndOfBody;
+              /* Commit subject  */ "%B" + EndOfBody;
```
Comment is now a little misleading. This is the body too, no?

gerhardol replied:
Will change to raw body

gerhardol replied:
Also updated a few other comments

RussKie commented:
We can now use Span<char> and ROS<char>, and avoid allocations.

On Mon, 31 May 2021, 9:14 am Drew Noakes, ***@***.***> wrote:

> ***@***.**** approved this pull request.
> ------------------------------
>
> In GitCommands/RevisionReader.cs
> <https://github.com/gitextensions/gitextensions/pull/9243#discussion_r642140551>
> :
>
> >              }
>
> -            string tail = reader.ReadToEnd() ?? "";
> -            int indexOfEndOfBody = tail.LastIndexOf(EndOfBody, StringComparison.InvariantCulture);
> +            // Ignore leading empty or whitespace lines, see Git pretty.c
> +            string tail = reader.ReadToEnd().TrimStart() ?? "";
>
> Is there a way to prevent allocating this tail string? Perhaps by
> extending the reader? For large repositories, this may create a lot of
> objects for the GC.
>
> Eg:
>
> reader.SkipWhiteSpace();
> var body = reader.ReadTo(EndOfBody);
> ------------------------------
>
> In GitCommands/RevisionReader.cs
> <https://github.com/gitextensions/gitextensions/pull/9243#discussion_r642141563>
> :
>
> > @@ -52,8 +52,7 @@ public sealed class RevisionReader : IDisposable
>                /* Author email    */ "%aE%n" +
>                /* Committer name  */ "%cN%n" +
>                /* Committer email */ "%cE%n" +
> -              /* Commit subject  */ "%s%n%n" +
> -              /* Commit body     */ "%b" + EndOfBody;
> +              /* Commit subject  */ "%B" + EndOfBody;
>
> Comment is now a little misleading. This is the body too, no?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9243#pullrequestreview-671872370>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXRHFN6KERMCBTSVYPLTQLBFPANCNFSM45XQ4XAQ>
> .
>

gerhardol commented:
Will add a few more changes to not allocate commit body that need to be garbage collected, maybe change endofbody too

gerhardol commented:
Preliminary improvements: Parsing about half time from before
I just ran a few times, noting "typical" times, no deep analysis.
It depends a little on when GC runs, should probably run the test at least 10 times.
But the difference is significant.

GE: 1.82-2.22s-> 0.99-1.12s
Linux: 12.0-13.0s -> 3.9-5.82s

mstv commented:
🎉 The loading times feel to have improved visibly. I will review in the next days.

gerhardol commented:
Two minor updates:

* Only store Body if different from Subject.
This makes no major difference for load times and memory usage. Maybe the string interning is helping here.
However, the code is clearer to me.
If this is not changed, the situations where Body is null (old commit where it is not stored) and MultiLine is false should be handled specially, no need to fetch the commit body.
See CommitInfo.cs:329, UpdateCommitMessage()

* Align use of %B vs %s%b and trimming subject/body
No real functional change

@mstv Can I squash before your review?

mstv commented:
> Can I squash before your review?

Yes, it is OK for me.

gerhardol commented:
Two first commits are same as #9253 
Last deals with other call sites and will be squashed
Previous commits are available in `tmp/log-zero-0`

mstv reviewed (commented):
Looks sensible.
A few existing NUnit test need to be adapted.

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -511,34 +466,82 @@ public StringLineReader(string s)
+
+            private void TrimEnd(ref int endIndex)
+            {
+                while (endIndex > _index && char.IsWhiteSpace(_s[endIndex - 1]))
```
This is a private method and endIndex is found in _s, so the check will not do anything.
(I made it private instead of a submethod to `ReadToEndOfBody()` to keep it next to TrimStart()

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -511,34 +466,82 @@ public StringLineReader(string s)
+                return (subject, body, hasMultiLineMessage);
+            }
+
+            public string? ReadToEnd()
+            {
+                if (_index >= _s.Length)
+                {
+                    return null;
+                }
+
+                return _s.Substring(_index);
```
I optimized that, but it is clearer, will change

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -33,27 +33,27 @@ public enum RefFilterOptions
     public sealed class RevisionReader : IDisposable
     {
-        private const string EndOfBody = "1DEA7CC4-FB39-450A-8DDF-762FCEA28B05";
+        private const char EndOfBody = (char)0x3;
```
additionalData is just filenames so the change should be safe
But the split should maybe be done in line 386 instead, encode additional privately (are evn filenames encoded here?)

mstv replied:
Unicode filenames could contain bytes with this value.
No need for a guid but perhaps `"\x03\x03\x03\x03"`.

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -511,34 +466,82 @@ public StringLineReader(string s)
+                    TrimEnd(ref endSubjectIndex);
+                }
+
+                // subject can also be defined as the contents before empty line (%s for --pretty),
+                // this uses the the alternative definition of first line in body.
```
ok

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -511,34 +466,82 @@ public StringLineReader(string s)
+                // subject can also be defined as the contents before empty line (%s for --pretty),
+                // this uses the the alternative definition of first line in body.
+                var subject = _s.Substring(_index, endSubjectIndex - _index);
+                var body = skipBody || endSubjectIndex == endBodyIndex
```
ok

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -511,34 +466,82 @@ public StringLineReader(string s)
+                var body = skipBody || endSubjectIndex == endBodyIndex
+                    ? null
+                    : _s.Substring(_index, endBodyIndex - _index);
+                bool hasMultiLineMessage = endSubjectIndex < endBodyIndex;
```
I had already changed that in a fix (I intended to change that earlier)

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -403,27 +392,21 @@ DateTime ParseUnixDateTime()
 
-            var subject = reader.ReadLine(advance: false);
+            // We keep full multiline message bodies within the last six months.
+            // Note also that if body and subject are identical (single line), the subject never need to be stored
```
ok

gerhardol commented on GitUI/UserControls/RevisionGrid/Columns/MessageColumnProvider.cs:
```diff
@@ -407,6 +409,7 @@ void DrawSuperProjectRef(string label, ref int currentOffset)
 
             if (lines.Length == 1)
             {
+                indicator.Render();
                 return;
             }
```
ok
Also in #9253

gerhardol commented on GitUI/UserControls/RevisionGrid/Columns/MessageColumnProvider.cs:
```diff
@@ -101,11 +101,13 @@ public override bool TryGetToolTip(DataGridViewCellMouseEventArgs e, GitRevision
-                var bodySummary = _gitRevisionSummaryBuilder.BuildSummary(revision.Body);
-                var initialLength = (bodySummary?.Length ?? 50) + 10;
+                // The body is not stored for older commits (to save memory)
+                string bodySummary = _gitRevisionSummaryBuilder.BuildSummary(revision.Body)
+                    ?? revision.Subject + (revision.HasMultiLineMessage ? TranslatedStrings.BodyNotLoaded : "");
+                int initialLength = bodySummary.Length + 10;
                 _toolTipBuilder.EnsureCapacity(initialLength);
 
-                _toolTipBuilder.Append(bodySummary ?? revision.Subject + TranslatedStrings.BodyNotLoaded);
+                _toolTipBuilder.Append(bodySummary);
```
I find this way better, will not change unless requested

mstv commented on GitCommands/Git/GitModule.cs:
```diff
@@ -950,11 +950,13 @@ public GitRevision GetRevision(ObjectId? objectId = null, bool shortFormat = fal
-                revision.Subject = revision.Body?.Substring(0, revision.Body.IndexOfAny(new[] { '\r', '\n' })) ?? "";
+                // See also RevisionReader for parsing commit body
+                revision.Body = ReEncodeCommitMessage(message, revision.MessageEncoding).TrimStart().TrimEnd();
+                int endSubjectIndex = revision.Body.IndexOf('\n');
+                revision.Subject = endSubjectIndex < 0 ? revision.Body : revision.Body.Substring(0, endSubjectIndex).TrimEnd();
```
```suggestion
                int? endSubjectIndex = revision.Body?.IndexOf('\n');
                revision.Subject = endSubjectIndex is not >= 0 ? revision.Body : revision.Body.Substring(0, endSubjectIndex).TrimEnd();
```

gerhardol replied:
It seems like `ReEncodeCommitMessage()` is incorrectly annotated.
A few other changes are required too.

mstv commented on UnitTests/GitUI.Tests/UserControls/RevisionGrid/Graph/LaneInfoProviderTests.cs:
```diff
@@ -234,7 +234,7 @@ public void GetLaneInfo_should_display_the_subject_if_singleline_body_null()
             _laneNodeLocator.FindPrevNode(Arg.Any<int>(), Arg.Any<int>()).Returns(x => (_realCommitNode, isAtNode: false));
 
-            GetLaneInfo_should_display(_realCommitNode, suffix: _realCommitNode.GitRevision.Subject);
+            GetLaneInfo_should_display(_realCommitNode);
```
No, the `LaneInfoProvider` must be adapted.

gerhardol replied:
Implementation in `LaneInfoProvider()` is OK. 
The difference now is that even if Body is null, Subject is used if `HasMultiLineMessage == false`

mstv replied:
I will check myself when MS VS works again.

mstv replied:
You are right.

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -511,34 +466,86 @@ public StringLineReader(string s)
-                if (advance)
+            public string? ReadToEnd()
+            {
+                if (_index >= _s.Length)
                 {
-                    _index = _s.Length;
+                    return null;
                 }
 
-                return s;
+                _index = _s.Length;
+
+                return _s.Substring(_index);
```
The substring must be taken before setting `_index = _s.Length`. The latter must be done for the `return null`, too.

gerhardol replied:
One reason I removed the advance of index here.
Renaming to `PeakToEnd()`, reversing the advancing

mstv replied:
I agree, but `PeekToEnd`.

mstv reviewed (commented):
There seems to be a need for NUnit tests for the `StringLineReader`.

gerhardol commented:
> Unicode filenames could contain bytes with this value.
No need for a guid but perhaps "\x03\x03\x03\x03".

It turns out that using a string to separate is considerably slower than a char. The total command is more than twice as long if a string is searched instead of a char like `_s.LastIndexOf((char)3`
To fix this, the parsing is aware of if there is a path filter or not 
The performance seem toimprove some tenths of percent further by this.

> There seems to be a need for NUnit tests for the `StringLineReader`.

The test should be for `TryParseRevision()` so not the tests are useless again at next refactoring.

I did not intend to make a change on the internals and originally just adopted the tests (the old tests were incorrect though). But the benefit in ditching the old implementation and avoid allocations were so big that I could not resist. I hate writing new tests though...

gerhardol commented:
> It turns out that using a string to separate is considerably slower than a char. The total command is more than twice as long if a string is searched instead of a char like `_s.LastIndexOf((char)3`
> To fix this, the parsing is aware of if there is a path filter or not
> The performance seem toimprove some tenths of percent further by this.

@RussKie 3.5 is has less execution time and less memory usage also compared to the optimized version in this PR. (This PR is maybe 25% worse than 3.5, master 400% worse). 
So there is a major regression with .NET5, not just for `LastIndexOf()`.

I was looking into backporting the change to not add endOfBody marker also to 3.5 but there is no major reason...

RussKie commented on GitCommands/Git/GitModule.cs:
```diff
@@ -950,11 +950,15 @@ public GitRevision GetRevision(ObjectId? objectId = null, bool shortFormat = fal
-                revision.Body = ReEncodeCommitMessage(message, revision.MessageEncoding);
-                revision.Subject = revision.Body?.Substring(0, revision.Body.IndexOfAny(new[] { '\r', '\n' })) ?? "";
+                // See also RevisionReader for parsing commit body
+                revision.Body = ReEncodeCommitMessage(message, revision.MessageEncoding)?.TrimStart().TrimEnd();
```
Why not 
```suggestion
                revision.Body = ReEncodeCommitMessage(message, revision.MessageEncoding)?.Trim();
```

Also why not trim inside `ReEncodeCommitMessage`?

gerhardol replied:
That affects a number of other calls too. I believe this is OK, changing.
It though means that GE _never_ present commit body as entered by the user.
But it is better that handling is consistent, this is what this part of the PR is all about.

RussKie commented on GitCommands/Git/GitModule.cs:
```diff
@@ -950,11 +950,15 @@ public GitRevision GetRevision(ObjectId? objectId = null, bool shortFormat = fal
                 // commit message is not re-encoded by git when format is given
-                revision.Body = ReEncodeCommitMessage(message, revision.MessageEncoding);
-                revision.Subject = revision.Body?.Substring(0, revision.Body.IndexOfAny(new[] { '\r', '\n' })) ?? "";
+                // See also RevisionReader for parsing commit body
+                revision.Body = ReEncodeCommitMessage(message, revision.MessageEncoding)?.TrimStart().TrimEnd();
+                int endSubjectIndex = revision.Body?.IndexOf('\n') ?? -1;
+                revision.Subject = endSubjectIndex >= 0
+                    ? revision.Body.Substring(0, endSubjectIndex).TrimEnd()
+                    : revision.Body ?? "";
```
We now have access to `ReadOnlySpan<char>`, so it would be significantly more performant to use it and splice as necessary. E.g. pseudo code:
```suggestion
                // See also RevisionReader for parsing commit body
                string body = ReEncodeCommitMessage(message, revision.MessageEncoding); // don't return null, return "" instead
                revision.Body = body;
                
                ReadOnlySpan<char> span = body.AsSpan();
                int endSubjectIndex = span.IndexOf('\n');
                revision.Subject = endSubjectIndex >= 0
                    ? span.Substring(0, endSubjectIndex).TrimEnd()
                    : "";
```

gerhardol replied:
This is not performance critical, but can change.
Adopting slightly (for instance not changing `ReEncodeCommitMessage()` - yet at least)

RussKie commented on GitCommands/RevisionReader.cs:
```diff
@@ -511,34 +470,101 @@ public StringLineReader(string s)
             }
 
-            public string? ReadToEnd(bool advance = true)
+            public StringLineReader TrimStart()
```
This should be leveraging .NET implementations instead of our custom ones (e.g. `ROS<char>`).
/cc: @drewnoakes

gerhardol replied:
I implemented this in RevisionReader, see branch `tmp/log-zero-0` (tests are not adapted)
However, it seem to be slower. linux repo loads in 6-7 seconds instead of 5-6 seconds.

This could be due to the adaptions I made to StringPool. This implementation could maybe be replaced with https://docs.microsoft.com/en-us/windows/communitytoolkit/high-performance/stringpool

RussKie replied:
First of all I'd look at rewriting the parsing logic to use ROS<char>, and then benchmark it. Then would add `StringPool`, and benchmark it again.
I'm finding it difficult to believe that our current implementation is less performant than with the use of ROS, but I haven't studied the code deeply enough to be sure.

RussKie replied:
This optimisation work is somewhat orthogonal to this change though.

gerhardol replied:
ROS was implemented - one reason this PR had so many feature creeps...

RussKie commented on GitUI/CommandsDialogs/FormVerify.LostObject.cs:
```diff
@@ -112,7 +112,7 @@ private LostObject(LostObjectType objectType, string rawType, ObjectId objectId)
                         result.Author = module.ReEncodeStringFromLossless(logPatternMatch.Groups[1].Value);
                         string encodingName = logPatternMatch.Groups[2].Value;
-                        result.Subject = module.ReEncodeCommitMessage(logPatternMatch.Groups[3].Value, encodingName);
+                        result.Subject = module.ReEncodeCommitMessage(logPatternMatch.Groups[3].Value, encodingName) ?? "";
```
`ReEncodeCommitMessage` mustn't return `null`.

gerhardol replied:
The method was previously not annotated `string?` but returned `ReEncodeStringFromLossless()` that returns `string?`. Some calls assumed `string?`, other not.

Do you want to change the implementation in `ReEncodeCommitMessage()` to return `?? ""`?

RussKie replied:
I'd look at changing the implementation of `ReEncodeCommitMessage` to never return `null` if inputs aren't `null`. May be we should add an attribute [`NotNullIfNotNull`](https://docs.microsoft.com/dotnet/csharp/language-reference/attributes/nullable-analysis#specify-conditional-post-conditions-notnullwhen-maybenullwhen-and-notnullifnotnull)?

RussKie commented:
> @RussKie 3.5 is has less execution time and less memory usage also compared to the optimized version in this PR. (This PR is maybe 25% worse than 3.5, master 400% worse).
> So there is a major regression with .NET5, not just for `LastIndexOf()`.

If you think there's a regression please raise an issue in https://github.com/dotnet/runtime

gerhardol commented:
> > @RussKie 3.5 is has less execution time and less memory usage also compared to the optimized version in this PR. (This PR is maybe 25% worse than 3.5, master 400% worse).
> > So there is a major regression with .NET5, not just for `LastIndexOf()`.
> 
> If you think there's a regression please raise an issue in https://github.com/dotnet/runtime

I do not have a good testcase for the general problem.
`LastIndexOf()` is quite clear to me, but I assume that refering to this PR is still not good enough.

Will not report anything right now.

mstv commented:
There is a nasty "%" being appended (both by 128949f and 90a9040):

![grafik](https://user-images.githubusercontent.com/36601201/120940473-c94a4980-c71d-11eb-8ad4-7684e8e8feee.png)

vbjay commented:
End of body is needed.

https://github.com/gitextensions/gitextensions/issues/6139


On Mon, Jun 7, 2021, 3:53 AM Michael Seibt ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> UnitTests/GitUI.Tests/UserControls/RevisionGrid/Graph/LaneInfoProviderTests.cs
> <https://github.com/gitextensions/gitextensions/pull/9243#discussion_r646351666>
> :
>
> > -            GetLaneInfo_should_display(_realCommitNode, suffix: _realCommitNode.GitRevision.Subject);
> +            GetLaneInfo_should_display(_realCommitNode);
>
> You are right.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9243#discussion_r646351666>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3WDMLPOCD2KJSJTFV2LCLTRR3IXANCNFSM45XQ4XAQ>
> .
>

gerhardol commented:
> End of body is needed. #6139

EndOfBody is included with path filter (and could be included with --name-only too if that was in RevisionReader).
In the normal case it is not needed (and will affect performance also if the .NET5 issue i solved).

mstv commented:
FileHistory / Blame seems to be broken.

vbjay commented:
File history was my point. They provide file names at the end. End of body
IS required.

On Tue, Jun 8, 2021, 5:39 AM Michael Seibt ***@***.***> wrote:

> FileHistory / Blame seems to be broken.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9243#issuecomment-856621710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3WDMLE6D2IBK4MSKMYXB3TRXQMZANCNFSM45XQ4XAQ>
> .
>

gerhardol commented:
> There is a nasty "%" being appended (both by 128949f and 90a9040)

Fixed

>                    FileHistory / Blame seems to be broken.

Worked for me, but there are artifacts, the body was not parsed correctly.
I do not get (Last)IndexOf working with a string of unprintable right now, something else need to be done.
As EndOfBody is only used when there is a file filter, this is not performance critical, so I leave it at this for now.

I suggest that this is rebased and squashed again...

mstv commented:
> I suggest that this is rebased and squashed again...

Yes, please.

gerhardol commented:
> > I suggest that this is rebased and squashed again...
> 
> Yes, please.

Fixed EndOfBody in tests, not changed in the commit yesterday:

Summary of open items as I interpret it:

* mstv wants more tests
* RussKie wants to use span in RevisionReader pushed to tmp/log-zero-0. Seem to slow down though, maybe due to my adaption in StringPool. Can we use high-performance toolkit instead?
* RussKie wants to change ReEncodeCommitMessage

The .NET5 implementation is significantly slower than .NET4 also with these optimizations. As the changes makes sense to me regardless, I suggest that is a later investigation.
(It seems like very few people actively use master right now or this would have been reported.)
This piece of code is probably the most important to optimize, it is a large part of the time opening a repo.

gerhardol commented:
I have tried `Microsoft.Toolkit.HighPerformance` `StringPool`, pushed to `tmp/log-zero-0`.
It is impossible to say which handling is the most efficient, The time varies too much.
Also if a GC is forced prior to making a refresh in the repo, the time `TryParseRevision()` requires to parse a repo varies from 2.4s to 4.3s over 6 refreshes, so it is not possible to determine the efficiency unless a very large number of refreshes are done.

If something, the current implementation seem to have lower low values compared to using ReadOnlySpan and `Microsoft.Toolkit.HighPerformance`. ReadOnlySpan with current StringPool seem to be slowest.
.NET5 is much slower too.
ReadOnlySpan should be faster so going that way should be a change for the better in the long run.

@drewnoakes any thoughts about the optimizations?

drewnoakes commented:
Sorry @gerhardol it's hard to say without taking a deeper look and I don't have time to dedicate to this right now. I wonder if the variability comes from git itself. Perhaps we can exclude that by having a perf test that consumes ready made data, so we are only measuring our code. That is, pipe the output of git to a byte array in memory, then use that in GE for testing/benchmarking (rather than streaming from the git process). I would hope that would reduce the variability and make it easier to see where the overhead is on the GE side.

gerhardol commented:
> Sorry @gerhardol it's hard to say without taking a deeper look and I don't have time to dedicate to this right now. I wonder if the variability comes from git itself. Perhaps we can exclude that by having a perf test that consumes ready made data, so we are only measuring our code. That is, pipe the output of git to a byte array in memory, then use that in GE for testing/benchmarking (rather than streaming from the git process). I would hope that would reduce the variability and make it easier to see where the overhead is on the GE side.

The time intervall is for `TryParseRevision()` only (start/stop StopWatch), so Git is not running and other GE activity should have minimal impact (assuming `subject.OnNext()` is not doing too much async which may be incorrect will try to just ignore that).
It is likely GC that causes the variability, but I do not know anything else than to force a GC before parsing the first parsing.
Maybe Profiler 'ticks' is more accurate to estimate what is taking time, that will ignore the GC.

I have identified a few other tweaks using Performance profiler (inlining, time parsing etc). But that is not changing much.

RussKie commented:
> It is impossible to say which handling is the most efficient, The time varies too much.

The best way to measure is to write benchmarks with https://github.com/dotnet/BenchmarkDotNet/.

RussKie commented:
I'm too very busy with work, and struggling to dedicate enough time to this atm. We can take the current change as is (if it is stable), and look at optimisations in the future.

drewnoakes commented:
> It is likely GC that causes the variability

You could verify this with PerfView. Use it to "Collect" a trace while you run the test. You'll be able to see all GC events. PerfView is not very friendly to use, but it's very powerful.

gerhardol commented:
> We can take the current change as is (if it is stable), and look at optimisations in the future.

I find this stable (if the code runs without hitting error exits a testcase would not find much more - the test cases in master is incorrect too.
I will try to add new test cases but have failed so far.
A benchmark would likely give the same problem as the actual run, so not attempted.

Using Diagnostic Tools sampling gives less variance than only StopWatch for TryParseRevision.
That seem to confirm that the use of ReadOnlySpan and HighPerformance StringBuffer decreases execution time so I go with that.

Identified a change that improve 15-30% (for TryParseRevision() runtime): Instead of converting commit unix time to DateTime when parsing do it when reading info for the commits.
The creation of the DateTime objects is not that bad (it requires allocation though), it is the calculation of LocalTime.
Not too much changes in other modules.
This part could be separated, but merges will be a pain.

ReadOnlySpan makes a minor improvement, not statistically proven but it should be better.
Surprisingly, it seems like ROS in the core function is more effective than in StringLineReader

With the total changes, the performance in master with these optimizations is comparable to 3.5
(So 3.5 has a lot more potential...)

The major hotspot with around 1/3 of the parsing is to parse ObjectId. I do not see much improvement possible there.
The second hotspot with around 1/4 is decoding of the output, which is hard to avoid.
The third hotspot is the StringPool and creating strings in general.

The two first commits in the PR is now included in #9300 (as they have been merged in 3.5)

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -351,13 +344,12 @@ DateTime ParseUnixDateTime()
                 }
             }
-
             #endregion
```
Nit: restore empty line before `#endregion`

gerhardol replied:
checked the file

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -131,33 +135,19 @@ public sealed class RevisionReader : IDisposable
-                    if (TryParseRevision(module, chunk, stringPool, logOutputEncoding, out var revision))
+                    if (TryParseRevision(module, chunk, logOutputEncoding, sixMonths, hasPathFilter, out var revision))
                     {
-                        if (revisionPredicate is null || revisionPredicate(revision))
```
`revisionPredicate` is not used / applied.

gerhardol replied:
reapplied

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -246,34 +237,35 @@ private static bool TryParseRevision(GitModule module, ArraySegment<byte> chunk,
-            var offset = chunk.Offset + (ObjectId.Sha1CharCount * 2);
-            var lastOffset = chunk.Offset + chunk.Count;
+            var offset = ObjectId.Sha1CharCount * 2;
+            var lastOffset = chunk.Count;
```
Should be renamed to `endOffset` in a separate commit.

gerhardol replied:
why separate?
applied now

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -338,10 +330,11 @@ int CountParents(int baseOffset)
             // Lines 2 and 3 are timestamps, as decimal ASCII seconds since the unix epoch, each terminated by `\n`
-            var authorDate = ParseUnixDateTime();
-            var commitDate = ParseUnixDateTime();
+            int startIndex = offset;
```
`startIndex` is not used, is it?

gerhardol replied:
removed
(I had implemented comparing the time slices to optimize, removed it to keep complexity down but missed this)

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -33,28 +34,29 @@ public enum RefFilterOptions
+            /* Commit raw body */ "%B";
 
+        // Separate commit message body and filtered file names
+        private const string EndOfBody = "1DEA7CC4-FB39-450A-8DDF-762FCEA28B05";
```
OK

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -246,34 +237,35 @@ private static bool TryParseRevision(GitModule module, ArraySegment<byte> chunk,
             // decoding as a string.
 
-            if (chunk.Count == 0)
+            ReadOnlySpan<byte> array = chunk.AsSpan();
```
OK

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -395,19 +395,23 @@ long ParseUnixDateTime(ref ReadOnlySpan<byte> array)
+            if (sixMonths > authorUnixTime || !hasMultiLineMessage)
+            {
+                body = null;
+            }
```
Why have you switched back to GC'ing the `body`?

gerhardol replied:
It This was to simplify, but I will revert so my previous version...

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -476,57 +480,41 @@ public StringLineReader(ReadOnlySpan<char> s)
-                    ? null
-                    : _s.Slice(_index, lengthBody).ToString();
+                int lengthSubject = bodySlice.IndexOf('\n');
+                bool hasMultiLineMessage = lengthSubject > 0;
```
❓
```suggestion
                bool hasMultiLineMessage = lengthSubject >= 0;
```

gerhardol replied:
Yes, formally better

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -395,23 +398,21 @@ long ParseUnixDateTime(ref ReadOnlySpan<byte> array)
+            (string? subject, string? body, bool hasMultiLineMessage) = reader.ReadToEndOfBody(hasPathFilter, skipBody);
+
+            // We keep a full multiline message body within the last six months.
+            // Note also that if body and subject are identical (single line), the subject never need to be stored
```
```suggestion
            // Note also that if body and subject are identical (single line), the body never need to be stored
```

gerhardol replied:
OK
(Also added additional comment on line 520)

mstv commented on GitCommands/RevisionReader.cs:
```diff
@@ -395,23 +398,21 @@ long ParseUnixDateTime(ref ReadOnlySpan<byte> array)
+            skipBody = skipBody || !hasMultiLineMessage;
 
-            if (author is null || authorEmail is null || committer is null || committerEmail is null || subject is null || body is null)
+            if (author is null || authorEmail is null || committer is null || committerEmail is null || subject is null || (!skipBody && body is null))
```
❓
```suggestion
            if (author is null || authorEmail is null || committer is null || committerEmail is null || subject is null || (skipBody != (body is null)))
```

gerhardol replied:
This is on the limit for pattern also for me, but changing as I assume I will prefer this in some time...

mstv reviewed (commented):
Yet another two nits

mstv commented:
👍 Seems to work well for me. Rebase will solve the false translation issue. Feel free to squash.

gerhardol commented:
> 👍 Seems to work well for me. Rebase will solve the false translation issue. Feel free to squash.

Kept the storing of dates separate, squashed the other changes.
(Two lines changed, cleanups.)

I am struggling with creating tests for TryParseRevision, see if I succeed.

RussKie commented:
Ready to merge?

gerhardol commented:
> Ready to merge?

I am writing tests...
(So basically, not being productive at all the last week.)
But I got a grip on how to do it yesterday, will add something.

gerhardol commented:
A few minor tweaks, preparations for testability and some tests for TryParseRevision()
Some special care for this section, this is probably the most important to optimize for performance.
This involves minimizing calling methods.

For the testability, I would have preferred to put TryParseRevision() in an internal class. It would add a class method call but skip adding a few variables on the stack so the same performance.
To split RevisionReader.cs is not a good option as the parsing must follow the arguments. It could be possible to add  the variables as private variables but the scope should be limited.
I find the current implementation the best compromise.

The tests are not great but relative complete. The code was rewritten some to make the tests pass in teorethic failures (but the flow is improved).
However, the tests fail on AppVeyor due to carrige return issues. 
I cannot figure out why, the tests look the same as similar Approval tests.
(The Json serialization of the objects is strange too, like that the actual sha in `ObjectId` is not included...)
Worst case, the tests could be ignored in AppVeyor.
(I have tried updating ApprovalTests.NET to the latest version, still fails.)

I have not done a formal benchmark.

The parsing when there is a pathFilter is not great (like parsing errors must be ignored), there are some other ways to achieve what we are after. But this is for a future PR, the parsing itself should be relative untouched by this.

gerhardol commented:
Will investigate #9315 or a followup before continuing this PR.
(It is complete, but simplifications should maybe be done prior to this PR.)

gerhardol commented:
Rebased on #9317 (required some cleanup) and a few squashed other fixups (see `tmp/log-zero-0` for details).
Tests are fixed (format and LocalTime complicates unit tests).
#9317 should be merged soon.

gerhardol commented:
@msftbot merge in 48 hours

gerhardol commented on UnitTests/GitCommands.Tests/RevisionReaderTests.cs:
```diff
@@ -78,70 +93,59 @@ public void BuildArguments_check_parameters(RefFilterOptions refFilterOptions, s
+        [TestCase("normal", true)]
+        [TestCase("short_sha", false)]
+        [TestCase("simple_pathfilter", true)]
+        [TestCase("subject_no_nody", true)]
```
done, 
squashed
